src/main/java/gregtech/common/gui/widget/orefilter/OreFilterTestSlot.java

```diff
@@ -0,0 +1,198 @@
+                this.testResult = new Object2BooleanAVLTreeMap<>();
+                success = false;
+                for (String oreDict : oreDicts) {
+                    boolean matches = glob != null && glob.matches(oreDict);
+                    this.testResult.put(oreDict, matches);
+                    success |= matches;
+                }
+                this.matchType = MatchType.ORE_DICT_MATCH;
+            }
+            updateAndNotifyMatchSuccess(this.expectedResult == success);
+            this.match.setVisible(this.expectedResult == success);
+            this.noMatch.setVisible(this.expectedResult != success);
+            return;
+        }
+        this.testResult = Object2BooleanMaps.emptyMap();
+        this.matchType = MatchType.INVALID;
+        updateAndNotifyMatchSuccess(false);
+        this.match.setVisible(false);
+        this.noMatch.setVisible(false);
+    }
+
+    private void updateAndNotifyMatchSuccess(boolean newValue) {
+        if (this.matchSuccess == newValue) return;
+        this.matchSuccess = newValue;
+        if (this.onMatchChange != null) {
+            this.onMatchChange.apply(newValue);
+        }
+    }
+
+    /**
+     * Get each test candidate for current state of test slot. An empty collection indicates that the
+     * match is for items without any ore dictionary entry. A {@code null} value indicates that the
+     * input state is invalid or empty.
+     *
+     * @return each test candidate for current state of test slot
+     */
+    @Nullable
+    protected abstract Set<String> getTestCandidates();
+
+    @Override
+    @SideOnly(Side.CLIENT)
+    public void drawInBackground(int mouseX, int mouseY, float partialTicks, IRenderContext context) {
+        Position pos = getPosition();
+        if (this.slotIcon != null) {
+            this.slotIcon.draw(pos.x, pos.y, 18, 18);
+        }
+
+        renderSlotContents(partialTicks, context);
+
+        if (isActive() && isMouseOverElement(mouseX, mouseY)) {
+            GlStateManager.disableDepth();
+            GlStateManager.colorMask(true, true, true, false);
+            drawSolidRect(getPosition().x + 1, getPosition().y + 1, 16, 16, 0x80ffffff);
+            GlStateManager.colorMask(true, true, true, true);
+            GlStateManager.enableBlend();
+        }
+
+        GlStateManager.disableDepth();
+        super.drawInBackground(mouseX, mouseY, partialTicks, context);
+        GlStateManager.enableDepth();
+    }
+
+    protected abstract void renderSlotContents(float partialTicks, IRenderContext context);
+
+    @Override
+    public void drawInForeground(int mouseX, int mouseY) {
+        if (isActive() && isMouseOverElement(mouseX, mouseY)) {
+            List<String> list;
+            switch (this.matchType) {
+                case NO_ORE_DICT_MATCH:
+                    list = Collections.singletonList(I18n.format(this.matchSuccess ?
+                            "cover.ore_dictionary_filter.test_slot.no_oredict.matches" :
```

Comment:
As this is not tied just to cover anymore it would be good idea to change the langKeys here

Reply:
I think it's fine to leave the lang keys as-is. The only other place the key would be reused in other context is on the tutorial app (which doesn't exist for now), and I think `cover.ore_dictionary_filter` still fits the usage.

Another benefit of using `cover.ore_dictionary_filter` is that it groups the other localization entries with same prefix, which should benefit translators.